imageio/core/imopen.py

```diff
@@ -0,0 +1,477 @@
+        """ Register a new plugin to be used when opening URIs.
+
+        Parameters
+        ----------
+        plugin_name : str
+            The name of the plugin to be registered. If the name already exists
+            it will be overwritten.
+        plugin_class : callable
+            A callable that returns an instance of a plugin that conforms
+            to the v3.0 API.
+        """
+
+        cls._known_plugins[plugin_name] = plugin_class
+
+
+class Plugin(object):
+    def __init__(self, uri):
+        """ Instantiate a new Legacy Plugin
+
+        Parameters
+        ----------
+        uri : {str, pathlib.Path, bytes, file}
+            The resource to load the image from, e.g. a filename, pathlib.Path,
+            http address or file object, see the docs for more info.
+
+        """
+        self._uri = uri
+
+    def read(self, *, index=None, **kwargs):
+        """
+        Parses the given URI and creates a ndarray from it.
+
+        Parameters
+        ----------
+        index : {integer}
+            If the URI contains a list of ndimages return the index-th image. If
+            None, read all ndimages in the URI and attempt to stack them along
+            a new 0-th axis (equivalent to np.stack(imgs, axis=0))
+        kwargs : ...
+            To be replaced by the implementing plugin. Custom plugin arguments
+            go here.
+        """
+        raise NotImplementedError
+
+    def write(self, image, **kwargs):
+        """
+        Write an ndimage to the specified URI.
+
+        If the URI points to a file on the current host and the file does not
+        yet exist it will be created. If the file exists already, it will be
+        appended if possible; otherwise, it will be replaced.
+
+        Parameters
+        ----------
+        image : numpy.ndarray
+            The ndimage or list of ndimages to write.
+        kwargs : ...
+            To be replaced by the implementing plugin. Custom plugin arguments
+            go here.
+        """
+        raise IOError(
+            f"The {self.__name__} plugin can not write to the given URI. "
+            "Try using a different plugin."
+        )
+
+    def iter(self, **kwargs):
+        """
+        Iterate over a list of ndimages given by the URI
+
+        Parameters
+        ----------
+        kwargs : ... To be replaced by the
+            implementing plugin. Custom plugin arguments go here.
+        """
+
+        raise NotImplementedError
+
+    def get_meta(self, *, index=None):
+        """ Read ndimage metadata from the URI
+
+        Parameters
+        ----------
+        index : {integer, None}
+            If the URI contains a list of ndimages return the metadata
+            corresponding to the index-th image. If None, behavior is decided
+            by the plugin. Typically, it returns the metadata for the 0-th
+            ndimage.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def can_open(cls, uri):
```

Comment:
Also have a look at this: https://github.com/imageio/imageio/issues/562

The proposed solution there is to try to instantiate the plugin/format (whatever its called) and it will raise an exception if it cannot deliver. This also removes the need for classmethods (which I find a bit ugly TBH :))

Reply:
I'll have a look at the linked issue and see if I can work it in :)

I'm a bit hesitant about the idea of tracking file extensions. A file may have the wrong ending or no ending at all, and the only way to identify it is via its content. We can still do caching based on file type to speed up routing once it becomes a bottleneck; however, I think we should keep the option to let a plugin decide what it can or cannot open and allow it to check more than just the extension if it sees fit.

Another consideration about having a class attribute with supported extensions is that we will run in periodic issues, where a plugin does support a format, but `imageio` doesn't because the tag is missing in the class attribute (#604, #586, #596, etc.). I'd prefer us not to track supported formats, which seems more smooth when upstream upgrades and supports new formats.

Reply:
But that would mean someone would have to write: `imageio.imwrite("aa.jpg", "PIL-jpg")` instead of just `imageio.imwrite("aa.jpg")`, right? That does not sound good, because then the user would need knowledge of the plugins of imageio. I know that it sucks when a plugin misses an extension, but it's also easily fixed. I do agree that a user should always be able to override the used plugin.

Reply:
@almarklein Yes, you are right. Writing is the current sore tooth of this PR and the reason I haven't converted it out of draft status yet. I totally agree that `imwrite(uri, numpy_data, metadata=..., **kwargs)` would be elegant.

`imwrite` doesn't exist in the new API (though I think we should add it), so the current way of writing files is a little cumbersome:

```
with iio.imopen(uri, plugin="name_of_plugin") as file:
    file.write(image)
```

The main reason I dislike keeping a register of extensions is that reading the file may be much more comprehensive than decisions based on the extension alone, and wrapping what ever means a plugin provides to identify a file seems easier to maintain. From the perspective of writing, the extension may be a good (enough?) option to go by.

How about this: We change `can_open` to `can_read` and `can_write`, which each check for their respective ability in the best way the plugin is able to. Upon reading/writing we pass the URI to each registered plugin in order and pick the first that doesn't raise an exception (unless `plugin=...` is specified by the user). This would mean adding a second mandatory argument to `imopen` in the form of `imopen(uri, mode, *, plugin=None, **kwargs)` where `mode` takes a value of `[r, w, rw]` (could be extended with modes like `a`, `r+`, etc.). We'd also have to come up with a better name than `mode` to not block this name as a kwarg for plugins; e.g. `io_mode`.

Thoughts?

Reply:
>  reading the file may be much more comprehensive than decisions based on the extension alone

Totally agree. But when a user wants to read from a file, you don't want to try all plugins until you find one that works. It would simply be too slow. If the file ends with a certain extension, we can at least start trying plugins that are associated with that extension (and the move to other plugins). In 95% of cases you'd have a direct hit.

I definitely think we need separate paths to test a plugins ability to read and write a file.

Reply:
> we can at least start trying plugins that are associated with that extension (and the move to other plugins)

@almarklein This is a nice suggestion. I would, however, wait with implementing this until we have enough plugins to actually see an impact on speed. If we can get away with just blindly trying them one-by-one, then this is much easier than thinking through edge cases of a lookup, e.g., how to handle multiple (non-equivalent) plugins for the same extension (gif can be written by us natively, or it can be written by pillow for example). Do you think we can move this into the next PR for now?

> I definitely think we need separate paths to test a plugins ability to read and write a file.

Done. I've introduced `io_mode` into `imopen` to allow selecting for either read (r) or write (w). I've also added an `imwrite` function. Does this look like a reasonable solution?